Get BlockingConnection into consistent state upon loss of TCP/IP connection with broker + acceptance tests

vitaly-krugl opened the pull request.

vitaly-krugl commented:
@gmr, this PR fixes a BlockingConnection regression introduced by df7d3b742bb8f9ef830454f82d4679bb0f219dca as well as a few other cases; it also adds several acceptance tests for loss of connection scenarios.

vitaly-krugl commented:
Fixed some issues and squashed all changes

vitaly-krugl commented on tests/acceptance/forward_server.py:
```diff
@@ -0,0 +1,409 @@
+"""TCP/IP forwarding/echo service for testing."""
```
@gmr, I pulled the code in this module from my own https://github.com/vitaly-krugl/inetpy repo that presently uses an "MIT License". Do I need to put any sort of copyright header in this file? Thanks, Vitaly

gmr replied:
Yeah, I'd just include the first paragraph bits in the top comment of the file.

wjps commented:
@vitaly-krugl, just seen a test failure of one of these tests on one of my travis builds and I think it's more likely the tests than the code. Care to take a look?

Full logs are here

https://travis-ci.org/wjps/pika/jobs/63049021

Stack trace...
```
BlockingConnection to downed broker results in AMQPConnectionError ... ok
BlockingConnection TCP/IP connection loss in CONNECTION_PROTOCOL ... ok
BlockingConnection TCP/IP connection loss in CONNECTION_START ... ERROR
BlockingConnection TCP/IP connection loss in CONNECTION_TUNE ... ok
BlockingConnection reconnect with downed broker ... ok
BlockingConnection resets properly on TCP/IP drop during channel() ... ok
======================================================================
ERROR: BlockingConnection TCP/IP connection loss in CONNECTION_START
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/wjps/pika/tests/acceptance/blocking_adapter_test.py", line 155, in start_test
    self.connection.connect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 234, in connect
    self._adapter_connect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 310, in _adapter_connect
    self.process_data_events()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 242, in process_data_events
    if self._handle_read():
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 358, in _handle_read
    super(BlockingConnection, self)._handle_read()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 389, in _handle_read
    self._on_data_available(data)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1334, in _on_data_available
    self._process_frame(frame_value)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1401, in _process_frame
    if self._process_callbacks(frame_value):
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1371, in _process_callbacks
    frame_value)  # Args
  File "/home/travis/build/wjps/pika/pika/callback.py", line 62, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/travis/build/wjps/pika/pika/callback.py", line 94, in wrapper
    return function(*args, **kwargs)
  File "/home/travis/build/wjps/pika/pika/callback.py", line 238, in process
    callback(*args, **keywords)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1319, in _on_connection_tune
    self._send_connection_open()
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1504, in _send_connection_open
    self._on_connection_open, [spec.Connection.OpenOk])
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1488, in _rpc
    self._send_method(channel_number, method_frame)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1554, in _send_method
    self._send_frame(frame.Method(channel_number, method_frame))
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 423, in _send_frame
    super(BlockingConnection, self)._send_frame(frame_value)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1539, in _send_frame
    self._flush_outbound()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 377, in _flush_outbound
    if self._handle_write():
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 417, in _handle_write
    return self._handle_error(error)
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 323, in _handle_error
    self._handle_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 273, in _handle_disconnect
    self._adapter_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 325, in _adapter_disconnect
    self._check_state_on_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 160, in _check_state_on_disconnect
    raise exceptions.ProbableAccessDeniedError
ProbableAccessDeniedError: 
-------------------- >> begin captured logging << --------------------
...
```

vitaly-krugl commented:
@wjps, thanks for the report. Is it reproducible on your travis setup? It seems to be passing on the official pika setup. I see that the tests were running at the latest cd8c9b08e8a121acfb8b53344292d7b2ff4eecc6. I will take a look after work.

wjps commented:
No it was a one off failure, exactly the same push passed on the official
pika setup minutes later. I suspect there's either another timing related
issue in there or something environmental. Like an idiot I've just told
travis to re-run the tests now, not realising it would trash the old logs,
sorry. It was getting 'connection refused' on a high port number though if
that's any use.





On Mon, 18 May 2015 at 19:44 vitaly-krugl <notifications@github.com> wrote:

> @wjps <https://github.com/wjps>, thanks for the report. Is it
> reproducible on your travis setup? It seems to be passing on the official
> pika setup. I see that the tests were running at the latest cd8c9b0
> <https://github.com/pika/pika/commit/cd8c9b08e8a121acfb8b53344292d7b2ff4eecc6>.
> I will take a look after work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pika/pika/pull/557#issuecomment-103168345>.
>

wjps commented:
Having said that I can reproduce it in a few dozen iterations with

```
$ true; while [ $? == 0 ]; do  nosetests -x -verbosity=3 tests/acceptance/blocking_adapter_test.py; done
```

vitaly-krugl commented:
I pulled up the log from my browser's history:
```
BlockingConnection TCP/IP connection loss in CONNECTION_PROTOCOL ... ok
BlockingConnection TCP/IP connection loss in CONNECTION_START ... ERROR
BlockingConnection TCP/IP connection loss in CONNECTION_TUNE ... ok
BlockingConnection reconnect with downed broker ... ok
BlockingConnection resets properly on TCP/IP drop during channel() ... ok
======================================================================
ERROR: BlockingConnection TCP/IP connection loss in CONNECTION_START
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/wjps/pika/tests/acceptance/blocking_adapter_test.py", line 155, in start_test
    self.connection.connect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 234, in connect
    self._adapter_connect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 310, in _adapter_connect
    self.process_data_events()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 242, in process_data_events
    if self._handle_read():
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 358, in _handle_read
    super(BlockingConnection, self)._handle_read()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 389, in _handle_read
    self._on_data_available(data)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1334, in _on_data_available
    self._process_frame(frame_value)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1401, in _process_frame
    if self._process_callbacks(frame_value):
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1371, in _process_callbacks
    frame_value)  # Args
  File "/home/travis/build/wjps/pika/pika/callback.py", line 62, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/travis/build/wjps/pika/pika/callback.py", line 94, in wrapper
    return function(*args, **kwargs)
  File "/home/travis/build/wjps/pika/pika/callback.py", line 238, in process
    callback(*args, **keywords)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1319, in _on_connection_tune
    self._send_connection_open()
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1504, in _send_connection_open
    self._on_connection_open, [spec.Connection.OpenOk])
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1488, in _rpc
    self._send_method(channel_number, method_frame)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1554, in _send_method
    self._send_frame(frame.Method(channel_number, method_frame))
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 423, in _send_frame
    super(BlockingConnection, self)._send_frame(frame_value)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1539, in _send_frame
    self._flush_outbound()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 377, in _flush_outbound
    if self._handle_write():
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 417, in _handle_write
    return self._handle_error(error)
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 323, in _handle_error
    self._handle_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 273, in _handle_disconnect
    self._adapter_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 325, in _adapter_disconnect
    self._check_state_on_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 160, in _check_state_on_disconnect
    raise exceptions.ProbableAccessDeniedError
ProbableAccessDeniedError: 
-------------------- >> begin captured logging << --------------------
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_error of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': False, 'arguments': None}
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.adapters.base_connection: INFO: Connecting to ::1:58952
pika.adapters.base_connection: WARNING: Connection to ::1:58952 failed: [Errno 111] Connection refused
pika.adapters.base_connection: INFO: Connecting to 127.0.0.1:58952
pika.callback: DEBUG: Processing 0:Connection.Start
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 0 registered uses left
pika.callback: DEBUG: Removing callback #0: {'callback': <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
pika.callback: DEBUG: Calling <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.Start"
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.callback: DEBUG: Processing 0:Connection.Tune
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 0 registered uses left
pika.callback: DEBUG: Removing callback #0: {'callback': <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
pika.callback: DEBUG: Calling <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.Tune"
pika.connection: DEBUG: Creating a HeartbeatChecker: 580
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_open of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.callback: DEBUG: Processing 0:Connection.OpenOk
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 0 registered uses left
pika.callback: DEBUG: Removing callback #0: {'callback': <bound method BlockingConnection._on_connection_open of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
pika.callback: DEBUG: Calling <bound method BlockingConnection._on_connection_open of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.OpenOk"
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_closed of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.adapters.blocking_connection: INFO: Closing connection (200): Normal shutdown
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_closed of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.callback: DEBUG: Processing 0:Connection.CloseOk
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 0 registered uses left
pika.callback: DEBUG: Removing callback #0: {'callback': <bound method BlockingConnection._on_connection_closed of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
pika.callback: DEBUG: Calling <bound method BlockingConnection._on_connection_closed of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.CloseOk"
pika.heartbeat: DEBUG: Removing timeout for next heartbeat interval
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.callback: DEBUG: Incremented callback reference counter: {'callback': <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 2}
pika.callback: DEBUG: Added: {'callback': <function <lambda> at 0x7f13477da5f0>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.adapters.base_connection: INFO: Connecting to ::1:58952
pika.adapters.base_connection: WARNING: Connection to ::1:58952 failed: [Errno 111] Connection refused
pika.adapters.base_connection: INFO: Connecting to 127.0.0.1:58952
pika.callback: DEBUG: Processing 0:Connection.Start
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 1 registered uses left
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 0 registered uses left
pika.callback: DEBUG: Removing callback #1: {'callback': <function <lambda> at 0x7f13477da5f0>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
pika.callback: DEBUG: Calling <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.Start"
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.callback: DEBUG: Calling <function <lambda> at 0x7f13477da5f0> for "0:Connection.Start"
pika.callback: DEBUG: Processing 0:Connection.Tune
pika.callback: DEBUG: Processing use of oneshot callback
pika.callback: DEBUG: 0 registered uses left
pika.callback: DEBUG: Removing callback #0: {'callback': <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
pika.callback: DEBUG: Calling <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.Tune"
pika.connection: DEBUG: Creating a HeartbeatChecker: 580
pika.callback: DEBUG: Added: {'callback': <bound method BlockingConnection._on_connection_open of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
pika.adapters.base_connection: ERROR: Fatal Socket Error: error(32, 'Broken pipe')
pika.heartbeat: DEBUG: Removing timeout for next heartbeat interval
pika.adapters.base_connection: ERROR: Socket closed while tuning the connection indicating a probable permission error when accessing a virtual host
pika.callback: DEBUG: Incremented callback reference counter: {'callback': <bound method BlockingConnection._on_connection_start of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 2}
pika.callback: ERROR: Calling <bound method BlockingConnection._on_connection_tune of <pika.adapters.blocking_connection.BlockingConnection object at 0x7f134877a490>> for "0:Connection.Tune" failed
Traceback (most recent call last):
  File "/home/travis/build/wjps/pika/pika/callback.py", line 238, in process
    callback(*args, **keywords)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1319, in _on_connection_tune
    self._send_connection_open()
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1504, in _send_connection_open
    self._on_connection_open, [spec.Connection.OpenOk])
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1488, in _rpc
    self._send_method(channel_number, method_frame)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1554, in _send_method
    self._send_frame(frame.Method(channel_number, method_frame))
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 423, in _send_frame
    super(BlockingConnection, self)._send_frame(frame_value)
  File "/home/travis/build/wjps/pika/pika/connection.py", line 1539, in _send_frame
    self._flush_outbound()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 377, in _flush_outbound
    if self._handle_write():
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 417, in _handle_write
    return self._handle_error(error)
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 323, in _handle_error
    self._handle_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 273, in _handle_disconnect
    self._adapter_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/blocking_connection.py", line 325, in _adapter_disconnect
    self._check_state_on_disconnect()
  File "/home/travis/build/wjps/pika/pika/adapters/base_connection.py", line 160, in _check_state_on_disconnect
    raise exceptions.ProbableAccessDeniedError
ProbableAccessDeniedError
--------------------- >> end captured logging << ---------------------
Name                                Stmts   Miss Branch BrMiss  Cover   Missing
-------------------------------------------------------------------------------
pika                                   20      4      0      0    80%   7-12
pika.adapters                          20      7      0      0    65%   32-33, 38-40, 44-45
pika.adapters.asyncore_connection      64     10      8      5    79%   51, 67-72, 98, 138-139
pika.adapters.base_connection         234     67     87     32    69%   16-17, 62, 117-120, 164, 191-192, 217-224, 232-245, 256, 259-261, 294, 297-298, 302-303, 309-320, 329, 341-342, 353-355, 358-359, 365, 369-381, 408-410, 414-415, 454
pika.adapters.blocking_connection     438    197    149     97    50%   31, 45-47, 76-77, 90-92, 144, 154, 167, 201, 217, 219, 243, 245, 253, 285-288, 337-338, 348, 392-394, 406, 409-411, 426-428, 490-495, 510-516, 539-582, 612, 626, 637-643, 655-667, 677-694, 720-728, 766, 786-787, 824-830, 845-846, 867-868, 894-897, 926-929, 947-948, 959-960, 978-980, 986-987, 994-999, 1003, 1007, 1015, 1020-1021, 1044-1045, 1054-1055, 1066-1067, 1070-1071, 1074-1084, 1092-1093, 1104-1105, 1108-1111, 1123-1130, 1133-1135, 1155, 1161, 1170-1172, 1194-1196, 1207, 1219
pika.adapters.libev_connection        122     28     38     15    73%   97, 128, 133, 145, 147, 153, 162, 164, 166, 176-177, 184-185, 211-214, 229-230, 234-245, 250-251
pika.adapters.select_connection       237     66     85     41    67%   74-76, 107-109, 158-163, 180-182, 215-216, 230, 309-316, 330-332, 360, 382-383, 393-416, 424-429, 438-451, 513-516
pika.adapters.tornado_connection       24      0      4      1    96%   
pika.adapters.twisted_connection      193    132     44     43    26%   31-32, 35-37, 40-42, 45-48, 70-75, 79-89, 98-111, 118-122, 129-131, 140-170, 173-174, 177-178, 183-185, 211, 219, 223-226, 230-233, 239-240, 246-253, 295, 300-301, 308, 317-318, 325-327, 332, 335, 339-342, 345, 348-349, 370-371, 382, 388, 392, 402-407, 422-424, 430, 434-436, 439-440, 444, 449-451, 454-458
pika.amqp_object                       25      0      6      0   100%   
pika.callback                         172      3     86      3    98%   224, 273-274
pika.channel                          342      9    124      9    96%   67, 300, 622, 716, 827-828, 908, 1097, 1136
pika.connection                       571     47    235     42    89%   13, 96, 135, 150, 180, 182, 197, 261, 504, 642, 653, 687, 764, 853, 862, 912, 973, 1013-1019, 1028-1033, 1037-1038, 1047, 1069, 1140, 1178, 1204-1205, 1281, 1351, 1395, 1409, 1432-1434, 1477, 1482, 1533-1534, 1541, 1588, 1608
pika.credentials                       28      0      6      0   100%   
pika.data                             144     21     68     10    85%   25, 135-136, 140-141, 145-146, 150-151, 160-161, 165-166, 175-176, 180-181, 193-196
pika.exceptions                        83     24     12      7    67%   7, 15, 21, 29, 42, 49, 56, 62-66, 72, 78-82, 88, 95, 101, 107, 113, 119, 133, 139, 145, 159
pika.frame                             89      0     14      0   100%   
pika.heartbeat                         65      0      8      1    99%   
pika.utils                              3      0      0      0   100%   
-------------------------------------------------------------------------------
TOTAL                                2874    615    974    306    76%   
----------------------------------------------------------------------
Ran 544 tests in 60.538s
FAILED (SKIP=16, errors=1)
The command "nosetests -c nose.cfg --with-coverage --cover-package=pika" exited with 1.
Done. Your build exited with 1.
```

vitaly-krugl commented:
@wjps, strangely, when running your loop on Fedora Linux (via VirtualBox) as well as natively on Mac OS X, I can reproduce a different exception that makes absolutely no sense (since I don't configure for SSL). Investigating. . .

```
======================================================================
ERROR: BlockingConnection reconnect with downed broker
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vkruglikov/ossdev/pika/tests/acceptance/blocking_adapter_test.py", line 109, in start_test
    channel = self.connection.channel()
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 200, in channel
    self._channels[channel_number] = BlockingChannel(self, channel_number)
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 473, in __init__
    self.open()
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 876, in open
    self._rpc(spec.Channel.Open(), self._on_openok, [spec.Channel.OpenOk])
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 1169, in _rpc
    self._wait_on_response(method_frame))
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 1190, in _send_method
    self.connection.process_data_events()
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 242, in process_data_events
    if self._handle_read():
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 358, in _handle_read
    super(BlockingConnection, self)._handle_read()
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/base_connection.py", line 373, in _handle_read
    except ssl.SSLWantReadError:
AttributeError: 'module' object has no attribute 'SSLWantReadError'
```

vitaly-krugl commented:
@wjps, it looks like you added `ssl.SSLWantReadError` exception handler in the "write starvation" PR.  It turns out that this exception was added in python 3.3, so it's not compatible with 2.x and is causing failures on 2.x:

Per https://docs.python.org/3/library/ssl.html:
```
exception ssl.SSLWantReadError
A subclass of SSLError raised by a non-blocking SSL socket when trying to read or write data, but more data needs to be received on the underlying TCP transport before the request can be fulfilled.

New in version 3.3.
```

wjps commented:
Hmm, I have ssl.WantReadError in my python 2.7.9,it must have been included
when SSL was backported for POODLE. It's not in < 2.7.9 though.

On Tue, 19 May 2015 at 06:39 vitaly-krugl <notifications@github.com> wrote:

> @wjps <https://github.com/wjps>, it looks like you added
> ssl.SSLWantReadError exception handler in the "write starvation" PR. It
> turns out that this exception was added in python 3.3, so it's not
> compatible with 2.x and is causing failures on 2.x:
>
> Per https://docs.python.org/3/library/ssl.html:
>
> exception ssl.SSLWantReadError
> A subclass of SSLError raised by a non-blocking SSL socket when trying to read or write data, but more data needs to be received on the underlying TCP transport before the request can be fulfilled.
>
> New in version 3.3.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pika/pika/pull/557#issuecomment-103345405>.
>

vitaly-krugl commented:
I have 2.7.6 on my Mac and recent installation of Fedora Linux came with 2.7.8.
